docs/design/2022-03-03-rc-read-tso-optimization.md

```diff
@@ -0,0 +1,36 @@
+# Read-Consistency Read With Timestamp Check
+
+- Author(s): cfzjywxk
+- Last updated: March. 3, 2022
+- Discussion at:
+
+## Motivation
+
+For the read-consistency isolation level read requests in a single transaction, each will need to fetch a new `ts` to read the latest committed data.
+If the workload is a read-heavy one or the read qps is large, fetching ts each time will increase the query lantecy.
+
+The new ts itself is used to ensure the most recent data will be returned, if the data version does not change frequently then it's unnecessary to fetch a new timestamp every time.
+That is the `rc-read` could be processed in an optimistic way, the ts could be updated only when a new version data is met, then the ts cost will be saved a lot for this case.
+
+## Detailed Design
+
+Introduce another system variable `tidb_rc_read_check_ts` to enable the "lazy ts update" read mode. It will take effect for all the in-transaction select stataments and the
+transaction isolation level is `read-consistency`.
+
+The execution flow is like this:
+
+1. If the query is executed first time, do not fetch a newer `for_update_ts` and just use the last valid one(`start_ts` for the first time).
+2. Do build the plan and executor like before.
+3. If the execution is successful, a `resultSet` will be returned.
+4. The connection layer will drive the response and call `Next` function using the `resultSet`.
+5. The read executor will try to fetch data to fill in a `chunk`. The bottommost executor could be a `pointGet`, `batchPointGet` or coprocessor task. What's different is a `RcReadCheckTS` flag will be set on the read requests.
+6. The read executor in storage layer will check the read results, if more recent version does exist then a `WriteConflict` error will be returned.
+7. For data write record, do check if there's new version compared with current `read_ts`.
+8. For lock record, return `ErrKeyIsLocked` though the `lock.ts` is greater than the `read_ts` as the `read_ts` could be a stale one.
+9. If no error is returned the query is finished. Otherwise if there's no `chunk` responsed to client yet, do retry the whole query and this time a new `for_update_ts` will be fetched.
```

Comment:
I think since tidb has to read it twice, why not read it the first time with maxint64, and then, calculate the tso brought back in the read response as a max+1 and read it again. This should avoid getting tso in RC?

Reply:
The timestamp in the read response is not certainly new enough. It is possible that there is a write record with larger commit TS in the third shard. And if the error is caused by a lock, we actually cannot know its commit TS.

Getting a new ts from TSO works simple and correct.

The only thing I might do different here is to get ts parallelly with the first read. I believe it won't increase much load thanks to the tso batching.